// @distributedlab/tools/dist/esm/duration.js
function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) {
        throw new TypeError("Cannot initialize the same private elements twice on an object");
    }
}
function _class_apply_descriptor_get(receiver, descriptor) {
    if (descriptor.get) {
        return descriptor.get.call(receiver);
    }
    return descriptor.value;
}
function _class_apply_descriptor_set(receiver, descriptor, value) {
    if (descriptor.set) {
        descriptor.set.call(receiver, value);
    } else {
        if (!descriptor.writable) {
            throw new TypeError("attempted to set read only private field");
        }
        descriptor.value = value;
    }
}
function _class_extract_field_descriptor(receiver, privateMap, action) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to " + action + " private field on non-instance");
    }
    return privateMap.get(receiver);
}
function _class_private_field_get(receiver, privateMap) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "get");
    return _class_apply_descriptor_get(receiver, descriptor);
}
function _class_private_field_init(obj, privateMap, value) {
    _check_private_redeclaration(obj, privateMap);
    privateMap.set(obj, value);
}
function _class_private_field_set(receiver, privateMap, value) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "set");
    _class_apply_descriptor_set(receiver, descriptor, value);
    return value;
}
import dayjs from "dayjs";
import durationlib from "dayjs/plugin/duration";
dayjs.extend(durationlib);
var _duration = new WeakMap();
export class Duration {
    _duration(input, unit) {
        if (typeof input == 'number') return dayjs.duration(input, unit);
        if (typeof input == 'string') return dayjs.duration(input);
        return dayjs.duration(input);
    }
    get duration() {
        return _class_private_field_get(this, _duration);
    }
    get asMilliseconds() {
        return _class_private_field_get(this, _duration).asMilliseconds();
    }
    get milliseconds() {
        return _class_private_field_get(this, _duration).milliseconds();
    }
    get asSeconds() {
        return _class_private_field_get(this, _duration).asSeconds();
    }
    get seconds() {
        return _class_private_field_get(this, _duration).seconds();
    }
    get asMinutes() {
        return _class_private_field_get(this, _duration).asMinutes();
    }
    get minutes() {
        return _class_private_field_get(this, _duration).minutes();
    }
    get asHours() {
        return _class_private_field_get(this, _duration).asHours();
    }
    get hours() {
        return _class_private_field_get(this, _duration).hours();
    }
    get asDays() {
        return _class_private_field_get(this, _duration).asDays();
    }
    get days() {
        return _class_private_field_get(this, _duration).days();
    }
    get asWeeks() {
        return _class_private_field_get(this, _duration).asWeeks();
    }
    get weeks() {
        return _class_private_field_get(this, _duration).weeks();
    }
    get asMonths() {
        return _class_private_field_get(this, _duration).asMonths();
    }
    get months() {
        return _class_private_field_get(this, _duration).months();
    }
    get asYears() {
        return _class_private_field_get(this, _duration).asYears();
    }
    get years() {
        return _class_private_field_get(this, _duration).years();
    }
    get humanized() {
        return _class_private_field_get(this, _duration).humanize();
    }
    format(formatStr) {
        return _class_private_field_get(this, _duration).format(formatStr);
    }
    constructor(input, unit){
        _class_private_field_init(this, _duration, {
            writable: true,
            value: void 0
        });
        _class_private_field_set(this, _duration, this._duration(input, unit));
    }
}
export function duration(input, unit) {
    return new Duration(input, unit);
} /*#__PURE__*/
