// @distributedlab/tools/dist/esm/event-emitter.js
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
export class EventEmitter {
    on(key, fn) {
        var _this;
        this.handlers[key] = (_this = this.handlers[key] || []) === null || _this === void 0 ? void 0 : _this.concat(fn);
    }
    once(key, fn) {
        const handler = (data)=>{
            fn(data);
            this.off(key, handler);
        };
        this.on(key, handler);
    }
    off(key, fn) {
        var _this;
        this.handlers[key] = (_this = this.handlers[key] || []) === null || _this === void 0 ? void 0 : _this.filter((f)=>f !== fn);
    }
    emit(key, data) {
        var _this;
        (_this = this.handlers[key] || []) === null || _this === void 0 ? void 0 : _this.forEach((fn)=>{
            fn(data);
        });
    }
    clear() {
        this.handlers = {};
    }
    constructor(){
        _define_property(this, "handlers", {});
    }
}
