// @distributedlab/w3p/dist/cjs/enums/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    CHAIN_TYPES: function() {
        return CHAIN_TYPES;
    },
    NEAR_CHAINS: function() {
        return NEAR_CHAINS;
    },
    SOLANA_CHAINS: function() {
        return SOLANA_CHAINS;
    }
});
var CHAIN_TYPES;
(function(CHAIN_TYPES) {
    CHAIN_TYPES["EVM"] = "EVM";
    CHAIN_TYPES["Solana"] = "Solana";
    CHAIN_TYPES["Near"] = "Near";
})(CHAIN_TYPES || (CHAIN_TYPES = {}));
var SOLANA_CHAINS;
(function(SOLANA_CHAINS) {
    SOLANA_CHAINS["DevNet"] = "devnet";
    SOLANA_CHAINS["TestNet"] = "testnet";
    SOLANA_CHAINS["MainNet"] = "mainnet";
})(SOLANA_CHAINS || (SOLANA_CHAINS = {}));
var NEAR_CHAINS;
(function(NEAR_CHAINS) {
    NEAR_CHAINS["TestNet"] = "testnet";
    NEAR_CHAINS["MainNet"] = "mainnet";
})(NEAR_CHAINS || (NEAR_CHAINS = {}));
