// @distributedlab/w3p/dist/cjs/enums/providers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    PROVIDERS: function() {
        return PROVIDERS;
    },
    PROVIDER_CHECKS: function() {
        return PROVIDER_CHECKS;
    },
    PROVIDER_EVENTS: function() {
        return PROVIDER_EVENTS;
    },
    PROVIDER_EVENT_BUS_EVENTS: function() {
        return PROVIDER_EVENT_BUS_EVENTS;
    }
});
var PROVIDERS;
(function(PROVIDERS) {
    PROVIDERS["Fallback"] = "fallback";
    PROVIDERS["Metamask"] = "metamask";
    PROVIDERS["Phantom"] = "phantom";
    PROVIDERS["Coinbase"] = "coinbase";
    PROVIDERS["Solflare"] = "solflare";
    PROVIDERS["Near"] = "near";
    PROVIDERS["WalletConnect"] = "wallet-connect";
})(PROVIDERS || (PROVIDERS = {}));
var PROVIDER_CHECKS;
(function(PROVIDER_CHECKS) {
    PROVIDER_CHECKS["Fallback"] = "isWeb3";
    PROVIDER_CHECKS["Metamask"] = "isMetaMask";
    PROVIDER_CHECKS["Coinbase"] = "isCoinbaseWallet";
    PROVIDER_CHECKS["Phantom"] = "isPhantom";
    PROVIDER_CHECKS["Solflare"] = "isSolflare";
    PROVIDER_CHECKS["Near"] = "isNear";
})(PROVIDER_CHECKS || (PROVIDER_CHECKS = {}));
var PROVIDER_EVENTS;
(function(PROVIDER_EVENTS) {
    PROVIDER_EVENTS["Connect"] = "connect";
    PROVIDER_EVENTS["Disconnect"] = "disconnect";
    PROVIDER_EVENTS["ChainChanged"] = "chainChanged";
    PROVIDER_EVENTS["AccountsChanged"] = "accountsChanged";
    PROVIDER_EVENTS["AccountChanged"] = "accountChanged";
})(PROVIDER_EVENTS || (PROVIDER_EVENTS = {}));
var PROVIDER_EVENT_BUS_EVENTS;
(function(PROVIDER_EVENT_BUS_EVENTS) {
    PROVIDER_EVENT_BUS_EVENTS["Connect"] = "connect";
    PROVIDER_EVENT_BUS_EVENTS["Disconnect"] = "disconnect";
    PROVIDER_EVENT_BUS_EVENTS["ChainChanged"] = "chain-changed";
    PROVIDER_EVENT_BUS_EVENTS["AccountChanged"] = "account-changed";
    PROVIDER_EVENT_BUS_EVENTS["Initiated"] = "initiated";
    PROVIDER_EVENT_BUS_EVENTS["BeforeTxSent"] = "before-tx-sent";
    PROVIDER_EVENT_BUS_EVENTS["TxSent"] = "tx-sent";
    PROVIDER_EVENT_BUS_EVENTS["TxConfirmed"] = "tx-confirmed";
    PROVIDER_EVENT_BUS_EVENTS["WalletConnectDisplayUri"] = "wallet-connect-display-uri";
})(PROVIDER_EVENT_BUS_EVENTS || (PROVIDER_EVENT_BUS_EVENTS = {}));
