// @distributedlab/w3p/dist/cjs/providers/raw/near-raw-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "NearRawProvider", {
    enumerable: true,
    get: function() {
        return NearRawProvider;
    }
});
const _core = require("@near-wallet-selector/core");
const _mynearwallet = require("@near-wallet-selector/my-near-wallet");
const _enums = require("../../enums");
const _types = require("../../types");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
let NearRawProvider = class NearRawProvider {
    init() {
        var _this = this;
        return _async_to_generator(function*() {
            _this.selector = yield (0, _core.setupWalletSelector)({
                network: _enums.NEAR_CHAINS.TestNet,
                modules: [
                    (0, _mynearwallet.setupMyNearWallet)()
                ]
            });
            const isSignedIn = _this.selector.isSignedIn();
            if (isSignedIn) {
                var _this_selector_store_getState_accounts_, _this_selector_store_getState_accounts, _this_selector_store_getState, _this_selector_store, _this_selector;
                _this.wallet = yield _this.selector.wallet();
                var _this_selector_store_getState_accounts__accountId;
                _this.accountId = (_this_selector_store_getState_accounts__accountId = (_this_selector = _this.selector) === null || _this_selector === void 0 ? void 0 : (_this_selector_store = _this_selector.store) === null || _this_selector_store === void 0 ? void 0 : (_this_selector_store_getState = _this_selector_store.getState()) === null || _this_selector_store_getState === void 0 ? void 0 : (_this_selector_store_getState_accounts = _this_selector_store_getState.accounts) === null || _this_selector_store_getState_accounts === void 0 ? void 0 : (_this_selector_store_getState_accounts_ = _this_selector_store_getState_accounts[0]) === null || _this_selector_store_getState_accounts_ === void 0 ? void 0 : _this_selector_store_getState_accounts_.accountId) !== null && _this_selector_store_getState_accounts__accountId !== void 0 ? _this_selector_store_getState_accounts__accountId : '';
            }
            return isSignedIn;
        })();
    }
    signIn() {
        var _this = this;
        return _async_to_generator(function*() {
            if (!_this.selector || Boolean(_this.accountId)) return;
            _this.wallet = yield _this.selector.wallet(_types.ENearWalletId.MyNearWallet);
            yield _this.wallet.signIn({
                contractId: _this.createAccessKeyFor,
                methodNames: [],
                accounts: []
            });
        })();
    }
    signOut() {
        var _this = this;
        return _async_to_generator(function*() {
            if (!_this.wallet) return;
            yield _this.wallet.signOut();
            _this.wallet = null;
            _this.accountId = '';
        })();
    }
    signAndSendTxs(txBody) {
        var _this = this;
        return _async_to_generator(function*() {
            if (!_this.wallet) return [];
            const outcome = yield _this.wallet.signAndSendTransactions(txBody);
            return outcome || [];
        })();
    }
    constructor({ createAccessKeyFor = '' }){
        _define_property(this, "selector", null);
        _define_property(this, "wallet", null);
        _define_property(this, "createAccessKeyFor", void 0);
        _define_property(this, "accountId", '');
        _define_property(this, "isNear", true);
        this.createAccessKeyFor = createAccessKeyFor;
    }
};
