// @distributedlab/tools/dist/cjs/duration.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Duration: function() {
        return Duration;
    },
    duration: function() {
        return duration;
    }
});
const _dayjs = _interop_require_default(require("dayjs"));
const _duration = _interop_require_default(require("dayjs/plugin/duration"));
function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) {
        throw new TypeError("Cannot initialize the same private elements twice on an object");
    }
}
function _class_apply_descriptor_get(receiver, descriptor) {
    if (descriptor.get) {
        return descriptor.get.call(receiver);
    }
    return descriptor.value;
}
function _class_apply_descriptor_set(receiver, descriptor, value) {
    if (descriptor.set) {
        descriptor.set.call(receiver, value);
    } else {
        if (!descriptor.writable) {
            throw new TypeError("attempted to set read only private field");
        }
        descriptor.value = value;
    }
}
function _class_extract_field_descriptor(receiver, privateMap, action) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to " + action + " private field on non-instance");
    }
    return privateMap.get(receiver);
}
function _class_private_field_get(receiver, privateMap) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "get");
    return _class_apply_descriptor_get(receiver, descriptor);
}
function _class_private_field_init(obj, privateMap, value) {
    _check_private_redeclaration(obj, privateMap);
    privateMap.set(obj, value);
}
function _class_private_field_set(receiver, privateMap, value) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "set");
    _class_apply_descriptor_set(receiver, descriptor, value);
    return value;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
_dayjs.default.extend(_duration.default);
var _duration1 = new WeakMap();
let Duration = class Duration {
    _duration(input, unit) {
        if (typeof input == 'number') return _dayjs.default.duration(input, unit);
        if (typeof input == 'string') return _dayjs.default.duration(input);
        return _dayjs.default.duration(input);
    }
    get duration() {
        return _class_private_field_get(this, _duration1);
    }
    get asMilliseconds() {
        return _class_private_field_get(this, _duration1).asMilliseconds();
    }
    get milliseconds() {
        return _class_private_field_get(this, _duration1).milliseconds();
    }
    get asSeconds() {
        return _class_private_field_get(this, _duration1).asSeconds();
    }
    get seconds() {
        return _class_private_field_get(this, _duration1).seconds();
    }
    get asMinutes() {
        return _class_private_field_get(this, _duration1).asMinutes();
    }
    get minutes() {
        return _class_private_field_get(this, _duration1).minutes();
    }
    get asHours() {
        return _class_private_field_get(this, _duration1).asHours();
    }
    get hours() {
        return _class_private_field_get(this, _duration1).hours();
    }
    get asDays() {
        return _class_private_field_get(this, _duration1).asDays();
    }
    get days() {
        return _class_private_field_get(this, _duration1).days();
    }
    get asWeeks() {
        return _class_private_field_get(this, _duration1).asWeeks();
    }
    get weeks() {
        return _class_private_field_get(this, _duration1).weeks();
    }
    get asMonths() {
        return _class_private_field_get(this, _duration1).asMonths();
    }
    get months() {
        return _class_private_field_get(this, _duration1).months();
    }
    get asYears() {
        return _class_private_field_get(this, _duration1).asYears();
    }
    get years() {
        return _class_private_field_get(this, _duration1).years();
    }
    get humanized() {
        return _class_private_field_get(this, _duration1).humanize();
    }
    format(formatStr) {
        return _class_private_field_get(this, _duration1).format(formatStr);
    }
    constructor(input, unit){
        _class_private_field_init(this, _duration1, {
            writable: true,
            value: void 0
        });
        _class_private_field_set(this, _duration1, this._duration(input, unit));
    }
};
function duration(input, unit) {
    return new Duration(input, unit);
} /*#__PURE__*/  /*#__PURE__*/  /*#__PURE__*/
