// @distributedlab/w3p/dist/esm/enums/rpc-error-codes.js
export var EIP1193;
(function(EIP1193) {
    EIP1193[EIP1193["UserRejectedRequest"] = 4001] = "UserRejectedRequest";
    EIP1193[EIP1193["Unauthorized"] = 4100] = "Unauthorized";
    EIP1193[EIP1193["UnsupportedMethod"] = 4200] = "UnsupportedMethod";
    EIP1193[EIP1193["Disconnected"] = 4900] = "Disconnected";
    EIP1193[EIP1193["ChainDisconnected"] = 4901] = "ChainDisconnected";
    EIP1193[EIP1193["UnrecognizedChain"] = 4902] = "UnrecognizedChain";
})(EIP1193 || (EIP1193 = {}));
export var EIP_1193_STRING;
(function(EIP_1193_STRING) {
    EIP_1193_STRING["UserRejectedRequest"] = "ACTION_REJECTED";
})(EIP_1193_STRING || (EIP_1193_STRING = {}));
export var EIP1474;
(function(EIP1474) {
    EIP1474[EIP1474["ParseError"] = -32700] = "ParseError";
    EIP1474[EIP1474["InvalidRequest"] = -32600] = "InvalidRequest";
    EIP1474[EIP1474["MethodNotFound"] = -32601] = "MethodNotFound";
    EIP1474[EIP1474["InvalidParams"] = -32602] = "InvalidParams";
    EIP1474[EIP1474["InternalError"] = -32603] = "InternalError";
    EIP1474[EIP1474["InvalidInput"] = -32000] = "InvalidInput";
    EIP1474[EIP1474["ResourceNotFound"] = -32001] = "ResourceNotFound";
    EIP1474[EIP1474["ResourceUnavailable"] = -32002] = "ResourceUnavailable";
    EIP1474[EIP1474["TransactionRejected"] = -32003] = "TransactionRejected";
    EIP1474[EIP1474["MethodNotSupported"] = -32004] = "MethodNotSupported";
    EIP1474[EIP1474["LimitExceeded"] = -32005] = "LimitExceeded";
    EIP1474[EIP1474["JsonRpcVersionNotSupported"] = -32006] = "JsonRpcVersionNotSupported";
})(EIP1474 || (EIP1474 = {}));
