// @distributedlab/w3p/dist/esm/providers/wrapped/_event-bus.js
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import { EventEmitter } from "@distributedlab/tools";
import { PROVIDER_EVENT_BUS_EVENTS } from "../../enums";
export class ProviderEventBus {
    emit(event, payload) {
        this.emitter.emit(event, payload);
    }
    onBeforeTxSent(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.BeforeTxSent, cb);
    }
    onTxSent(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.TxSent, cb);
    }
    onTxConfirmed(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.TxConfirmed, cb);
    }
    onInitiated(cb) {
        this.emitter.once(PROVIDER_EVENT_BUS_EVENTS.Initiated, cb);
    }
    onConnect(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.Connect, cb);
    }
    onDisconnect(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.Disconnect, cb);
    }
    onAccountChanged(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.AccountChanged, cb);
    }
    onChainChanged(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.ChainChanged, cb);
    }
    onWalletConnectDisplayUri(cb) {
        this.emitter.on(PROVIDER_EVENT_BUS_EVENTS.WalletConnectDisplayUri, cb);
    }
    clearHandlers() {
        this.emitter.clear();
    }
    constructor(){
        _define_property(this, "emitter", new EventEmitter());
    }
}
