// @distributedlab/w3p/dist/cjs/providers/wrapped/_base-evm.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseEVMProvider", {
    enumerable: true,
    get: function() {
        return BaseEVMProvider;
    }
});
const _ethers = require("ethers");
const _enums = require("../../enums");
const _amount = require("../../helpers/amount");
const _eth = require("../../helpers/eth");
const _eventbus = require("./_event-bus");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
let BaseEVMProvider = class BaseEVMProvider extends _eventbus.ProviderEventBus {
    get chainType() {
        return _enums.CHAIN_TYPES.EVM;
    }
    get isConnected() {
        return Boolean(this.chainId && this.address);
    }
    get defaultEventPayload() {
        return {
            address: this.address,
            chainId: this.chainId,
            isConnected: this.isConnected
        };
    }
    init() {
        var _this = this;
        return _async_to_generator(function*() {
            yield _this.setListeners();
            const currentAccounts = yield _this.provider.listAccounts();
            const network = yield _this.provider.getNetwork();
            _this.address = currentAccounts[0];
            _this.chainId = (0, _amount.hexToDecimal)(network.chainId);
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.Initiated, _this.defaultEventPayload);
        })();
    }
    switchChain(chainId) {
        var _this = this;
        return _async_to_generator(function*() {
            yield (0, _eth.requestSwitchEthChain)(_this.provider, chainId);
        })();
    }
    addChain(chain) {
        var _this = this;
        return _async_to_generator(function*() {
            yield (0, _eth.requestAddEthChain)(_this.provider, chain);
        })();
    }
    connect() {
        var _this = this;
        return _async_to_generator(function*() {
            yield (0, _eth.connectEthAccounts)(_this.provider);
        })();
    }
    getAddressUrl(chain, address) {
        return (0, _eth.getEthExplorerAddressUrl)(chain, address);
    }
    getTxUrl(chain, txHash) {
        return (0, _eth.getEthExplorerTxUrl)(chain, txHash);
    }
    getHashFromTx(txResponse) {
        return txResponse.transactionHash;
    }
    signAndSendTx(tx) {
        var _this = this;
        return _async_to_generator(function*() {
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.BeforeTxSent, {
                txBody: tx
            });
            const transactionResponse = yield _this.provider.getSigner().sendTransaction(tx);
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.TxSent, {
                txHash: transactionResponse.hash
            });
            const receipt = yield transactionResponse.wait();
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.TxConfirmed, {
                txResponse: receipt
            });
            return receipt;
        })();
    }
    signMessage(message) {
        var _this = this;
        return _async_to_generator(function*() {
            return _this.provider.getSigner().signMessage(message);
        })();
    }
    setListeners() {
        var _this = this;
        return _async_to_generator(function*() {
            const stubProvider = _this.provider.provider;
            stubProvider.on(_enums.PROVIDER_EVENTS.AccountsChanged, /*#__PURE__*/ _async_to_generator(function*() {
                const currentAccounts = yield _this.provider.listAccounts();
                var _currentAccounts_;
                _this.address = (_currentAccounts_ = currentAccounts[0]) !== null && _currentAccounts_ !== void 0 ? _currentAccounts_ : '';
                _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.AccountChanged, _this.defaultEventPayload);
                _this.emit(_this.isConnected ? _enums.PROVIDER_EVENT_BUS_EVENTS.Connect : _enums.PROVIDER_EVENT_BUS_EVENTS.Disconnect, _this.defaultEventPayload);
            }));
            stubProvider.on(_enums.PROVIDER_EVENTS.ChainChanged, (chainId)=>{
                _this.chainId = (0, _amount.hexToDecimal)(chainId);
                _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.ChainChanged, _this.defaultEventPayload);
            });
        })();
    }
    constructor(provider){
        super();
        _define_property(this, "provider", void 0);
        _define_property(this, "rawProvider", void 0);
        _define_property(this, "chainId", void 0);
        _define_property(this, "address", void 0);
        this.provider = new _ethers.providers.Web3Provider(provider, 'any');
        this.rawProvider = provider;
    }
};
