// @distributedlab/w3p/dist/cjs/providers/wrapped/_base-solana.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseSolanaProvider", {
    enumerable: true,
    get: function() {
        return BaseSolanaProvider;
    }
});
const _web3 = require("@solana/web3.js");
const _enums = require("../../enums");
const _solana = require("../../helpers/solana");
const _eventbus = require("./_event-bus");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
const getAddress = (publicKey)=>{
    return publicKey ? new _web3.PublicKey(publicKey).toBase58() : '';
};
let BaseSolanaProvider = class BaseSolanaProvider extends _eventbus.ProviderEventBus {
    get chainType() {
        return _enums.CHAIN_TYPES.Solana;
    }
    get isConnected() {
        return Boolean(this.chainId && this.address);
    }
    get defaultEventPayload() {
        return {
            chainId: this.chainId,
            address: this.address,
            isConnected: this.isConnected
        };
    }
    init() {
        var _this = this;
        return _async_to_generator(function*() {
            _this.setListeners();
            _this.address = getAddress(_this.provider.publicKey);
            _this.chainId = _enums.SOLANA_CHAINS.DevNet;
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.Initiated, _this.defaultEventPayload);
        })();
    }
    switchChain(chainId) {
        var _this = this;
        return _async_to_generator(function*() {
            try {
                _this.chainId = chainId;
                _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.ChainChanged, _this.defaultEventPayload);
            } catch (error) {
                (0, _solana.handleSolError)(error);
            }
        })();
    }
    connect() {
        var _this = this;
        return _async_to_generator(function*() {
            try {
                yield _this.provider.connect();
            } catch (error) {
                (0, _solana.handleSolError)(error);
            }
        })();
    }
    getAddressUrl(chain, address) {
        return (0, _solana.getSolExplorerAddressUrl)(chain, address);
    }
    getTxUrl(chain, txHash) {
        return (0, _solana.getSolExplorerTxUrl)(chain, txHash);
    }
    getHashFromTx(txResponse) {
        return txResponse;
    }
    signAndSendTx(// eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    txRequestBody) {
        return _async_to_generator(function*() {
            throw new TypeError('Method should be implemented in extender class');
        })();
    }
    setListeners() {
        this.provider.on(_enums.PROVIDER_EVENTS.Connect, ()=>{
            this.address = getAddress(this.provider.publicKey);
            this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.Connect, this.defaultEventPayload);
        });
        this.provider.on(_enums.PROVIDER_EVENTS.Disconnect, ()=>{
            this.address = getAddress(this.provider.publicKey);
            this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.Disconnect, this.defaultEventPayload);
        });
        this.provider.on(_enums.PROVIDER_EVENTS.AccountChanged, ()=>{
            this.address = getAddress(this.provider.publicKey);
            this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.AccountChanged, this.defaultEventPayload);
        });
    }
    constructor(provider){
        super();
        _define_property(this, "provider", void 0);
        _define_property(this, "rawProvider", void 0);
        _define_property(this, "chainId", void 0);
        _define_property(this, "address", void 0);
        this.provider = provider;
        this.rawProvider = provider;
    }
};
