// @distributedlab/w3p/dist/cjs/providers/wrapped/_event-bus.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ProviderEventBus", {
    enumerable: true,
    get: function() {
        return ProviderEventBus;
    }
});
const _tools = require("@distributedlab/tools");
const _enums = require("../../enums");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
let ProviderEventBus = class ProviderEventBus {
    emit(event, payload) {
        this.emitter.emit(event, payload);
    }
    onBeforeTxSent(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.BeforeTxSent, cb);
    }
    onTxSent(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.TxSent, cb);
    }
    onTxConfirmed(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.TxConfirmed, cb);
    }
    onInitiated(cb) {
        this.emitter.once(_enums.PROVIDER_EVENT_BUS_EVENTS.Initiated, cb);
    }
    onConnect(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.Connect, cb);
    }
    onDisconnect(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.Disconnect, cb);
    }
    onAccountChanged(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.AccountChanged, cb);
    }
    onChainChanged(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.ChainChanged, cb);
    }
    onWalletConnectDisplayUri(cb) {
        this.emitter.on(_enums.PROVIDER_EVENT_BUS_EVENTS.WalletConnectDisplayUri, cb);
    }
    clearHandlers() {
        this.emitter.clear();
    }
    constructor(){
        _define_property(this, "emitter", new _tools.EventEmitter());
    }
};
