// @distributedlab/w3p/dist/cjs/providers/wrapped/fallback-evm.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "FallbackEvmProvider", {
    enumerable: true,
    get: function() {
        return FallbackEvmProvider;
    }
});
const _enums = require("../../enums");
const _helpers = require("../../helpers");
const _eventbus = require("./_event-bus");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
let FallbackEvmProvider = class FallbackEvmProvider extends _eventbus.ProviderEventBus {
    static get providerType() {
        return _enums.PROVIDERS.Fallback;
    }
    get chainType() {
        return _enums.CHAIN_TYPES.EVM;
    }
    get isConnected() {
        return Boolean(this.chainId);
    }
    get defaultEventPayload() {
        return {
            address: this.address,
            chainId: this.chainId,
            isConnected: this.isConnected
        };
    }
    init() {
        var _this = this;
        return _async_to_generator(function*() {
            const network = yield _this.provider.getNetwork();
            _this.chainId = (0, _helpers.hexToDecimal)(network.chainId);
            _this.emit(_enums.PROVIDER_EVENT_BUS_EVENTS.Initiated, _this.defaultEventPayload);
        })();
    }
    getAddressUrl(chain, address) {
        return (0, _helpers.getEthExplorerAddressUrl)(chain, address);
    }
    getTxUrl(chain, txHash) {
        return (0, _helpers.getEthExplorerTxUrl)(chain, txHash);
    }
    getHashFromTx(txResponse) {
        return txResponse.transactionHash;
    }
    constructor(provider){
        super();
        _define_property(this, "provider", void 0);
        _define_property(this, "rawProvider", void 0);
        _define_property(this, "chainId", void 0);
        _define_property(this, "address", void 0);
        this.provider = provider;
        this.rawProvider = provider;
    }
};
