// @distributedlab/w3p/dist/cjs/helpers/near.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    MAX_GAS_LIMIT: function() {
        return MAX_GAS_LIMIT;
    },
    NO_DEPOSIT: function() {
        return NO_DEPOSIT;
    },
    getNearExplorerAddressUrl: function() {
        return getNearExplorerAddressUrl;
    },
    getNearExplorerTxUrl: function() {
        return getNearExplorerTxUrl;
    },
    handleNearError: function() {
        return handleNearError;
    },
    nearToYocto: function() {
        return nearToYocto;
    },
    yoctoToNear: function() {
        return yoctoToNear;
    }
});
const _nearapijs = require("near-api-js");
const _enums = require("../enums");
const _errors = require("../errors");
const MAX_GAS_LIMIT = '300000000000000';
const NO_DEPOSIT = '0';
const nearToYocto = (amount)=>{
    return _nearapijs.utils.format.parseNearAmount(amount);
};
const yoctoToNear = (amount)=>{
    return _nearapijs.utils.format.formatNearAmount(amount);
};
function getNearExplorerTxUrl(explorerUrl, txHash) {
    return `${explorerUrl}/transactions/${txHash}`;
}
function getNearExplorerAddressUrl(explorerUrl, address) {
    return `${explorerUrl}/accounts/${address}`;
}
function handleNearError(error) {
    switch(error.code){
        case _enums.EIP1193.UserRejectedRequest:
            throw new _errors.errors.ProviderUserRejectedRequest();
        case _enums.EIP1193.Unauthorized:
            throw new _errors.errors.ProviderUnauthorized();
        case _enums.EIP1193.UnsupportedMethod:
            throw new _errors.errors.ProviderUnsupportedMethod();
        case _enums.EIP1193.Disconnected:
            throw new _errors.errors.ProviderDisconnected();
        case _enums.EIP1193.ChainDisconnected:
            throw new _errors.errors.ProviderChainDisconnected();
        case _enums.EIP1474.ParseError:
            throw new _errors.errors.ProviderParseError();
        case _enums.EIP1474.InvalidRequest:
            throw new _errors.errors.ProviderInvalidRequest();
        case _enums.EIP1474.MethodNotFound:
            throw new _errors.errors.ProviderMethodNotFound();
        case _enums.EIP1474.InvalidParams:
            throw new _errors.errors.ProviderInvalidParams();
        case _enums.EIP1474.InternalError:
            throw new _errors.errors.ProviderInternalError();
        case _enums.EIP1474.InvalidInput:
            throw new _errors.errors.ProviderInvalidInput();
        case _enums.EIP1474.ResourceNotFound:
            throw new _errors.errors.ProviderResourceNotFound();
        case _enums.EIP1474.ResourceUnavailable:
            throw new _errors.errors.ProviderResourceUnavailable();
        case _enums.EIP1474.TransactionRejected:
            throw new _errors.errors.ProviderTransactionRejected();
        case _enums.EIP1474.MethodNotSupported:
            throw new _errors.errors.ProviderMethodNotSupported();
        case _enums.EIP1474.LimitExceeded:
            throw new _errors.errors.ProviderLimitExceeded();
        case _enums.EIP1474.JsonRpcVersionNotSupported:
            throw new _errors.errors.ProviderJsonRpcVersionNotSupported();
        default:
            throw error;
    }
}
